### PR TITLE
refactor(audio): split output.rs into render submodules

### PR DESCRIPTION
**Pull request by thedavidweng**
## Summary

`audio/output.rs` was 4,062 lines (about 2,580 of them inline tests). The module was already **deep** — its interface is just `ensure_output_thread`, `render_output_buffer` and `ResamplerCache` — but the flat file made navigation costly in a file that changed 30 times in four months.

Implementation is now partitioned into `output/{mix_bus,crossfade_render,cpal_stream,resampler_cache}.rs`, all `pub(super)`. The public interface in `mod.rs` is unchanged and no caller is edited. Tests moved alongside the submodule they exercise, verified line-by-line as an equivalent set.

Also fixes a real packaging hazard: the repo's `output/` build-artifact ignore rule silently swallowed the whole new `src-tauri/src/audio/output/` directory, so a negation entry is added to `.gitignore`. Without it the four new source files would never enter version control.

## Verification

`cargo check --all-targets` clean on this branch alone; `git check-ignore` confirms every new file is tracked. ADR-0005 (lock-free realtime callback) and ADR-0006 (all-or-nothing stem buffering) semantics are preserved verbatim — this is a pure move with no new locks or allocations on the render path.

## Summary by CodeRabbit

- **New Features**
  - Added a complete audio output pipeline with improved mixing, equal-power crossfades, resampling, EQ, limiting, peak tracking, and gapless transitions.
  - Added support for streaming and decoded audio stems with synchronized playback and channel mapping.
  - Added AirPlay forwarding, local-output suppression, and recovery when audio devices are disconnected.

- **Bug Fixes**
  - Improved handling of buffering, underruns, pauses, seeks, silence, and mismatched sample rates.
  - Preserved audio output tracking across crossfade transitions.

- **Chores**
  - Kept generated audio output files tracked for development and distribution.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -37,6 +37,7 @@ dist/
 build/
 out/
 output/
+!src-tauri/src/audio/output/
```
[the remark this answers is not in the document]
Comment (thedavidweng) [reply .gitignore:40]: Keeping the negation. The unanchored `output/` rule is deliberate and the exception is the conservative way to carve one path out of it.

Verified: with `main`'s `.gitignore`, `git check-ignore -v --no-index src-tauri/src/audio/output/mod.rs` reports `.gitignore:39:output/`. Without `!src-tauri/src/audio/output/`, the four new source files never enter version control at all — the split would land as a deleted `output.rs` and nothing else.

On anchoring instead: `/output/` would work today (nothing in the repo's tooling writes to a nested `output/`), but it is a broader change than it looks. It flips the ignore semantics for *every* nested `output/` path in the tree, not just the one this PR adds, and it would leave the surrounding block inconsistent — `dist/`, `build/`, `out/` and `coverage/` on lines 36-42 are all unanchored by the same convention. Re-anchoring that block is a repo-hygiene change that should stand on its own and be evaluated against the build outputs it governs, not ride along in an audio refactor.

File: src-tauri/src/audio/output/cpal_stream.rs [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+                thread::sleep(OUTPUT_DEVICE_RETRY_INTERVAL);
+                continue;
+            }
+        };
+        rebuild_failure_logged = false;
+
+        // Drop prepared/crossfade from the prior output-format generation.
+        if let Ok(mut controller) = playback.try_lock() {
+            controller.cancel_crossfade_and_prepared();
+        }
```
[the remark this answers is not in the document]
Comment (thedavidweng) [reply src-tauri/src/audio/output/cpal_stream.rs:171]: Pre-existing, not introduced here. This PR is a pure relocation of `src-tauri/src/audio/output.rs` into `src-tauri/src/audio/output/`: comparing the concatenated new modules against `main:src-tauri/src/audio/output.rs` line for line, the only differences are `super::` prefixes dropped from tests that now sit beside their subject, four new `#[cfg(test)]` module headers, reshuffled `use` statements, and two deleted section-banner comments. No statement changed.

`build_output_stream` is unchanged from `main:src-tauri/src/audio/output.rs` L1176-1278, `try_lock` included.

On the merits, the suggested swap is not obviously safe and needs its own change. Switching to a blocking `lock()` on the stream-rebuild path introduces a wait on a mutex that the CPAL device callback also takes; the current `try_lock` is what keeps rebuild from stalling behind a callback. And the proposed `Err(_) => return Err(...)` arm turns a poisoned controller into a hard rebuild failure, where today the stream still comes up. Changing the locking discipline between the rebuild path and the realtime callback is a concurrency change, so it deserves a dedicated PR with reasoning about the callback side, not a rider on a move.

File: src-tauri/src/audio/output/cpal_stream.rs [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+            let _ = tx.send(Ok(()));
+        }
+
+        while !shutdown.load(Ordering::Relaxed) && !device_lost.load(Ordering::SeqCst) {
+            thread::sleep(OUTPUT_DEVICE_POLL_INTERVAL);
+        }
+
+        if device_lost.load(Ordering::SeqCst) {
+            tracing::warn!("audio output device went away; rebuilding the stream");
+        }
```
[the remark this answers is not in the document]
Comment (thedavidweng) [reply src-tauri/src/audio/output/cpal_stream.rs:198]: Out of scope: this is a feature request, not a defect in the diff, and the code is unchanged by this PR.

The rebuild loop lives in `start_output_thread`, relocated verbatim from `main:src-tauri/src/audio/output.rs` L1279-1359. This PR is a pure move — the only differences anywhere in the split are dropped `super::` prefixes in tests, four `#[cfg(test)]` module headers, reshuffled imports, and two deleted banner comments.

Following the system default output device as it changes is a real product behavior with its own design questions (polling interval on the audio thread, interaction with the explicit device selection the app already exposes, and what should happen mid-crossfade). It belongs in its own change.

File: src-tauri/src/audio/output/cpal_stream.rs [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+    }
+
+    downmix_for_airplay_into(&scratch[..rendered_samples], channels, airplay_scratch);
+    if !airplay_scratch.is_empty() {
+        let owned = std::mem::replace(
+            airplay_scratch,
+            Vec::with_capacity(airplay_scratch.capacity()),
+        );
+        airplay_audio_tap.push_interleaved(sample_rate, 2, owned);
+    }
```
[the remark this answers is not in the document]
Comment (thedavidweng) [reply src-tauri/src/audio/output/cpal_stream.rs:286]: Pre-existing and unchanged by this PR. `forward_rendered_audio_to_airplay` is relocated verbatim from `main:src-tauri/src/audio/output.rs` L1417-1443. The split is a pure move: the only differences across all five new modules are `super::` prefixes dropped from tests, four `#[cfg(test)]` module headers, reshuffled imports, and two deleted banner comments.

Worth noting for whoever picks this up: the allocation is on the AirPlay forwarding path, which only runs when a route is actually engaged, so it is not on the unconditional hot path. Introducing a buffer pool means adding shared mutable state that the device callback touches, which needs to be designed against `AirPlayAudioTap`'s ownership model rather than bolted onto a relocation.

File: src-tauri/src/audio/output/crossfade_render.rs [hunk 5]
```diff
@@ -0,0 +1,1384 @@
...
+            active.rendered_frames += mix_frames as u64;
+            active.incoming_source_frame += inc_consumed;
+        }
+
+        chunk_start += chunk_frames;
+
+        if inc_rendered == 0 && mix_frames == 0 {
+            break;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (thedavidweng) [reply src-tauri/src/audio/output/crossfade_render.rs:146]: Flagging this one for separate follow-up rather than fixing it here, because it is pre-existing and the change it asks for is a redesign of the crossfade loop.

`render_crossfade_overlap` is relocated verbatim from `main:src-tauri/src/audio/output.rs` L517 onward. This PR is a pure move: comparing the concatenated new modules against `main`'s single `output.rs` line for line, the only differences are `super::` prefixes dropped from tests that now sit beside their subject, four `#[cfg(test)]` module headers, reshuffled `use` statements, and two deleted section-banner comments. No statement changed, so nothing in this diff can have introduced the divergence between `chunk_start`, `rendered_output_frames` and `active.rendered_frames`.

The suggested remedy — committing every cursor on a single frame count and adding overlap recovery when the outgoing track is exhausted — is explicitly a heavy lift, and it would change audible crossfade behavior. That needs its own PR with targeted tests for the unequal-resampler-output case; the existing crossfade tests in this module deliberately cover the equal-advance paths only. Landing it inside a relocation would make both the refactor and the behavior change unreviewable.

File: src-tauri/src/audio/output/mix_bus.rs [hunk 6]
```diff
@@ -0,0 +1,1022 @@
...
+    audio: &DecodedAudio,
+    start_frame: u64,
+    gain: f32,
+    device_sample_rate: u32,
+    device_channels: usize,
+    resampler_cache: Option<&mut ResamplerCache>,
+) -> (usize, u64) {
+    if gain == 0.0 {
+        return (0, 0);
+    }
```
[the remark this answers is not in the document]
Comment (thedavidweng) [reply src-tauri/src/audio/output/mix_bus.rs:17]: Pre-existing; not changed by this PR, and not something to fix inside a move.

`mix_stem_resampled` is relocated verbatim from `main:src-tauri/src/audio/output.rs` L703 onward. The split is a pure move — the only differences across the five new modules are `super::` prefixes dropped from tests, four `#[cfg(test)]` module headers, reshuffled imports, and two deleted banner comments.

The observation about the transport clock at `gain == 0.0` is worth its own investigation, and I am raising it separately. It is also not a one-line change: making the zero-gain path report consumed source frames alters what `advance_render_frame` sees on the decoded single-track path, the crossfade overlap accounting, and EOF/gapless-swap timing at volume 0. That needs tests that pin the transport position at zero volume, which do not exist today.

File: src-tauri/src/audio/output/mix_bus.rs [hunk 7]
```diff
@@ -0,0 +1,1022 @@
...
+
+fn mix_stem_same_rate(
+    output: &mut [f32],
+    audio: &DecodedAudio,
+    start_frame: u64,
+    gain: f32,
+    device_channels: usize,
+) -> (usize, u64) {
+    let src_channels = audio.channels;
+    let total_src_frames = audio.samples.len() / src_channels;
```
[the remark this answers is not in the document]
Comment (thedavidweng) [reply src-tauri/src/audio/output/mix_bus.rs:53]: Pre-existing and unchanged here. `mix_stem_same_rate` and its siblings are relocated verbatim from `main:src-tauri/src/audio/output.rs` (the `let total_src_frames = audio.samples.len() / src_channels;` divisions sit at L750, L790 and L848 there). This PR is a pure move — the only differences across the split are `super::` prefixes dropped from tests, four `#[cfg(test)]` module headers, reshuffled imports, and two deleted banner comments.

The consistency point is fair as a hardening item. It is worth noting it is defensive rather than a live defect: `channels` originates from decoded track metadata and from `AudioConsumer`, and a zero-channel track would already be malformed well before it reaches the mix bus. Adding `.max(1)` uniformly, or better, rejecting a zero-channel track at the boundary where it is constructed, is a standalone change.

File: src-tauri/src/audio/output/mix_bus.rs [hunk 8]
```diff
@@ -0,0 +1,1022 @@
...
+    let output_frames = output.len() / device_channels;
+
+    // Feed exactly input_frames_next(); zero-pad only at end-of-track.
+    let input_needed = resampler_cache
+        .get_or_create_mut(audio.sample_rate_hz, device_sample_rate, 0, output_frames)
+        .resampler
+        .input_frames_next();
+    let real_available = total_src_frames - src_start_frame;
+    let frames_from_source = real_available.min(input_needed);
+    let feed_frames = input_needed;
```
[the remark this answers is not in the document]
Comment (thedavidweng) [reply src-tauri/src/audio/output/mix_bus.rs:167]: Pre-existing; relocated without modification. `mix_stem_rubato` comes from `main:src-tauri/src/audio/output.rs` (the `let feed_frames = input_needed;` line is at L864 there) and `resample_mix_to_output` from L942. This PR is a pure move: the only differences across the five new modules are `super::` prefixes dropped from tests, four `#[cfg(test)]` module headers, reshuffled imports, and two deleted banner comments.

I am flagging the substance separately, since a channel that silently desynchronizes with no recovery is a real robustness gap. The fix you describe — evicting the whole cache entry on a per-channel `process` error so the next callback rebuilds from clean state — looks like the right shape, and it is a change to realtime resampler state that needs its own tests for the error path. That does not belong in a relocation.

File: src-tauri/src/audio/output/mod.rs [hunk 9]
```diff
@@ -0,0 +1,1241 @@
...
+                    }
+                    peak_accumulator.process(remaining, extra_rendered, device_channels, peak_ring);
+                    playback.advance_render_frame(extra_frames);
+                    total_rendered += extra_rendered;
+                }
+            }
+
+            return total_rendered;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (thedavidweng) [reply src-tauri/src/audio/output/mod.rs:248]: The premise is wrong on both counts, so I am not acting on this.

"The refactor copied two pipeline stages instead of extracting them" — it did not copy anything. This PR is a pure relocation of `src-tauri/src/audio/output.rs` into `src-tauri/src/audio/output/`. Comparing the concatenated new modules against `main:src-tauri/src/audio/output.rs` line for line, the only differences are `super::` prefixes dropped from tests that now sit beside their subject, four `#[cfg(test)]` module headers, reshuffled `use` statements, and two deleted section-banner comments. Both copies of both stages already exist on `main`, inside the single file:

- the crossfade-branch pipeline at `main` L253-311 and the main-path pipeline further down the same function
- the per-channel resampler feed in `mix_stem_rubato` (`main` L~838-940) and in `resample_mix_to_output` (`main` L942 onward)

"the missing fade gain at `mod.rs` lines 221-243 shows that this already failed once" — that asymmetry is also on `main`, verbatim, at L298-308. It is not evidence that this refactor drifted; nothing in this diff could have caused it.

The underlying suggestion (extract the shared stages) is reasonable as future work, and it is explicitly a heavy lift touching the realtime render path. Doing it here would mean this PR is no longer a relocation anyone can verify by inspection, which is the entire value of the change.

File: src-tauri/src/audio/output/mod.rs [hunk 10]
```diff
@@ -0,0 +1,1241 @@
...
+                        }
+                    } else {
+                        for sample in remaining[..extra_rendered].iter_mut() {
+                            *sample = soft_limit(*sample);
+                        }
+                    }
+                    peak_accumulator.process(remaining, extra_rendered, device_channels, peak_ring);
+                    playback.advance_render_frame(extra_frames);
+                    total_rendered += extra_rendered;
+                }
```
[the remark this answers is not in the document]
Comment (thedavidweng) [reply src-tauri/src/audio/output/mod.rs:243]: The asymmetry is real, but it is pre-existing and this PR did not introduce it — so I am flagging it for separate follow-up rather than fixing it inside a move.

The crossfade branch, including the post-swap tail that skips the fade gain, is byte-identical to `main:src-tauri/src/audio/output.rs` L253-311. This PR is a pure relocation: comparing the concatenated new modules against `main`'s single `output.rs` line for line, the only differences are `super::` prefixes dropped from tests, four `#[cfg(test)]` module headers, reshuffled `use` statements, and two deleted banner comments. No statement changed.

Two reasons not to take the patch here. First, applying a fade gain to post-swap samples changes audible output in the narrow window where a gapless swap and an active fade coincide during a crossfade render, and there is no test pinning that interaction today — the fix needs one. Second, this is exactly the kind of behavior change that makes a relocation unreviewable: the value of this PR is that a reviewer can confirm no statement moved semantically, and a "small" gain application in the middle of it destroys that property.

File: src-tauri/src/audio/output/resampler_cache.rs [hunk 11]
```diff
@@ -0,0 +1,89 @@
+use rubato::{Async, FixedAsync, SincInterpolationParameters, SincInterpolationType};
+use std::collections::HashMap;
+
+#[derive(Default)]
+pub struct ResamplerCache {
+    pub(super) cache: HashMap<(u32, u32, usize, usize), ResamplerEntry>,
+}
```
[the remark this answers is not in the document]
Comment (thedavidweng) [reply src-tauri/src/audio/output/resampler_cache.rs:7]: Pre-existing and unchanged. `ResamplerCache` is relocated verbatim from `main:src-tauri/src/audio/output.rs` L25-38. This PR is a pure move: the only differences across the five new modules are `super::` prefixes dropped from tests, four `#[cfg(test)]` module headers, reshuffled imports, and two deleted banner comments.

On the merits, bounding the cache is a plausible hardening item, but the growth is bounded in practice by the number of distinct callback frame counts a device actually produces, not by an unbounded key space, and evicting the wrong entry mid-stream discards sinc history and produces an audible discontinuity. Any eviction policy has to be designed against that, which makes it its own change rather than a rider on a relocation.

File: src-tauri/src/audio/output/resampler_cache.rs [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+fn resample_ratio(src_rate: u32, dst_rate: u32) -> f64 {
+    dst_rate as f64 / src_rate as f64
+}
```
[the remark this answers is not in the document]
Comment (thedavidweng) [reply src-tauri/src/audio/output/resampler_cache.rs:11]: Pre-existing; relocated without modification from `main:src-tauri/src/audio/output.rs` L25-38 (`resample_ratio` at L29, the `.expect` in `get_or_create_mut` further down the same impl). This PR is a pure move — the only differences across the split are `super::` prefixes dropped from tests, four `#[cfg(test)]` module headers, reshuffled imports, and two deleted banner comments.

I am raising the substance separately, because "panic on the realtime thread" is the right thing to care about. One correction to the reachability argument, though: a zero `src_rate` reaching the resampler would mean a track or consumer was constructed with no sample rate, which is a defect at the decode/streaming boundary rather than something the render callback should paper over. Threading an `Option` through `get_or_create_mut` and falling back to the linear path would hide that upstream defect as silent quality degradation. The better shape is probably to reject a zero sample rate where the track metadata is built, and only then decide whether the callback still needs a non-panicking path. Either way it is a design decision, not a mechanical edit.
